src/test/services/user.test.ts

```diff
@@ -0,0 +1,246 @@
+import { USER_ROLES, UserRole } from "../../IsaacApiTypes";
+import {
+  extractTeacherName,
+  isAdmin,
+  isAdminOrEventManager,
+  isEventLeader,
+  isEventLeaderOrStaff,
+  isEventManager,
+  isLoggedIn,
+  isStaff,
+  isStudent,
+  isTeacher,
+  isTeacherOrAbove,
+  isTeacherPending,
+  isTutor,
+  isTutorOrAbove,
+  roleRequirements,
+  schoolNameWithPostcode,
+} from "../../app/services/user";
+import { School } from "../../IsaacAppTypes";
+import { mockUser } from "../../mocks/data";
+
+const generateTestCases = (functionName: string) => {
+  const roleRequirements: Record<UserRole, string[]> = {
+    STUDENT: ["isStudent"],
+    TUTOR: ["isTutorOrAbove", "isTutor"],
+    TEACHER: ["isTeacherOrAbove", "isTeacher", "isTutorOrAbove"],
+    EVENT_LEADER: ["isEventLeader", "isTeacherOrAbove", "isTutorOrAbove", "isEventLeaderOrStaff"],
+    CONTENT_EDITOR: ["isStaff", "isTeacherOrAbove", "isTutorOrAbove", "isEventLeaderOrStaff"],
+    EVENT_MANAGER: [
+      "isEventManager",
+      "isStaff",
+      "isTeacherOrAbove",
+      "isTutorOrAbove",
+      "isEventLeaderOrStaff",
+      "isAdminOrEventManager",
+    ],
+    ADMIN: [
+      "isAdmin",
+      "isStaff",
+      "isTeacherOrAbove",
+      "isTutorOrAbove",
+      "isEventLeaderOrStaff",
+      "isAdminOrEventManager",
+    ],
+  };
+
+  const generateTestCase = (
+    role: UserRole,
+  ): {
+    role: UserRole | undefined | null;
+    value: { role: UserRole; loggedIn: boolean } | undefined | null;
+    expected: boolean;
+  } => {
+    const value = { role, loggedIn: true };
+    const expected = roleRequirements[role].includes(functionName);
+    return { role, value, expected };
+  };
+
+  const testCases = USER_ROLES.map(generateTestCase);
+  testCases.push({ role: undefined, value: undefined, expected: false }, { role: null, value: null, expected: false });
+
+  return testCases;
+};
+
+describe("User Checks", () => {
+  describe("isStudent function", () => {
+    const testCases = generateTestCases("isStudent");
```

Comment:
You can probably take this a step further and refactor these tests, since they're all the same. So maybe rather than passing the function name here, you can pass the function itself.

Reply:
I've changed it to pass the function itself, and now have a .forEach to do the same thing for each function.